test/distributed/_composable/fsdp/test_fully_shard_init.py

```diff
@@ -554,6 +566,91 @@ def test_invalid_meta_device_init(self):
         with self.assertRaisesRegex(RuntimeError, error_regex):
             model(inp)
 
+    @unittest.skipIf(not TEST_CUDA, "no cuda")
+    def test_rank0_broadcast_meta_device_init(self):
+        model_args = ModelArgs(dropout_p=0.0)
+        # Assume we have a CPU full state dict on rank 0
+        if self.rank == 0:
+            torch.manual_seed(42)
+            ref_model = Transformer(model_args)
+            full_sd = ref_model.state_dict()
+            for param in full_sd.values():
+                self.assertEqual(param.device, torch.device("cpu"))
+
+        # Initialize the sharded model on meta device
+        fsdp_mesh = init_device_mesh("cuda", (self.world_size,))
+        with torch.device("meta"):
+            model = Transformer(model_args)
+        for module in model.modules():
+            if isinstance(module, TransformerBlock):
+                fully_shard(module, mesh=fsdp_mesh)
+        fully_shard(model, mesh=fsdp_mesh)
+        for param in model.parameters():
+            self.assertEqual(param.device, torch.device("meta"))
+
+        # Construct a sharded state dict from the rank 0 full state dict by
+        # broadcasting and sharding
+        meta_sharded_sd = model.state_dict()
```

Comment:
The idea behind this conversion is that the since `DTensor`s encode the sharding info at the tensor level, we can use the sharded meta-device state dict to know how to shard the unsharded tensors from the full state dict (via `distribute_tensor()`).